rclrs/src/timer.rs

```diff
@@ -0,0 +1,812 @@
+///
+/// The public API of [`Timer`] is implemented via `TimerState`.
+///
+/// Timers that run inside of a [`Worker`] are represented by [`WorkerTimer`].
+pub struct TimerState<Scope: WorkScope> {
+    pub(crate) handle: Arc<TimerHandle>,
+    /// The callback function that runs when the timer is due.
+    callback: Mutex<Option<AnyTimerCallback<Scope>>>,
+    /// What was the last time lapse between calls to this timer
+    last_elapse: Mutex<Duration>,
+    /// We use Mutex<Option<>> here because we need to construct the TimerState object
+    /// before we can get the lifecycle handle.
+    #[allow(unused)]
+    lifecycle: Mutex<Option<WaitableLifecycle>>,
+    _ignore: std::marker::PhantomData<Scope>,
+}
+
+impl<Scope: WorkScope> TimerState<Scope> {
+    /// Gets the period of the timer
+    pub fn get_timer_period(&self) -> Result<Duration, RclrsError> {
+        let mut timer_period_ns = 0;
+        unsafe {
+            let rcl_timer = self.handle.rcl_timer.lock().unwrap();
+            rcl_timer_get_period(&*rcl_timer, &mut timer_period_ns)
+        }
+        .ok()?;
+
+        rcl_duration(timer_period_ns)
+    }
+
+    /// Cancels the timer, stopping the execution of the callback
+    pub fn cancel(&self) -> Result<(), RclrsError> {
+        let mut rcl_timer = self.handle.rcl_timer.lock().unwrap();
+        let cancel_result = unsafe { rcl_timer_cancel(&mut *rcl_timer) }.ok()?;
+        Ok(cancel_result)
+    }
+
+    /// Checks whether the timer is canceled or not
+    pub fn is_canceled(&self) -> Result<bool, RclrsError> {
+        let mut is_canceled = false;
+        unsafe {
+            let rcl_timer = self.handle.rcl_timer.lock().unwrap();
+            rcl_timer_is_canceled(&*rcl_timer, &mut is_canceled)
+        }
+        .ok()?;
+        Ok(is_canceled)
+    }
+
+    /// Get the last time lapse between calls to the timer.
+    ///
+    /// This is different from [`Self::time_since_last_call`] because it remains
+    /// constant between calls to the Timer.
+    ///
+    /// It keeps track of the what the value of [`Self::time_since_last_call`]
+    /// was immediately before the most recent call to the callback. This will
+    /// be [`Duration::ZERO`] if the `Timer` has never been triggered.
+    pub fn last_elapse(&self) -> Duration {
+        *self.last_elapse.lock().unwrap()
+    }
+
+    /// Retrieves the time since the last call to the callback
+    pub fn time_since_last_call(&self) -> Result<Duration, RclrsError> {
+        let mut time_value_ns: i64 = 0;
+        unsafe {
+            let rcl_timer = self.handle.rcl_timer.lock().unwrap();
+            rcl_timer_get_time_since_last_call(&*rcl_timer, &mut time_value_ns)
+        }
+        .ok()?;
+
+        rcl_duration(time_value_ns)
+    }
+
+    /// Retrieves the time until the next call of the callback
+    pub fn time_until_next_call(&self) -> Result<Duration, RclrsError> {
+        let mut time_value_ns: i64 = 0;
+        unsafe {
```

Comment:
Shouldn't this have a safety comment?

Reply:
[a589997](https://github.com/ros2-rust/ros2_rust/pull/480/commits/a589997b24b07b7a934b7431604f25d8e82a815a)